util/kernel_names.h

```diff
@@ -0,0 +1,139 @@
+
+struct adapter_byte {};
+struct adapter_size_t {};
+
+/** @brief Provides type that is safe to use in kernel names based on type given
+ *  @details There is a plenty of restrictions on types we may use for kernel
+ *           naming. The most obvious is non-forward-declarable classes, but
+ *           there are more of them.
+ *           This function uses discarded statements to map input type into the
+ *           return type we may use for the kernel name template parameters.
+ *           Note: it does not provide different kernel names for the type
+ *           aliases, so we may have the same kernel name
+ *             - for `int` and `int32_t`, or
+ *             - for `size_t` and `uint64_t`, or
+ *             - for any other possible type alias
+ *           on any specific SYCL implementation. We may say it provides adapter
+ *           aliases for the type aliases.
+ */
+template <typename T>
+constexpr auto adapter_helper() {
+  // Fixed-width types are optional and implementation-defined
+#ifdef INT8_MAX
+  if constexpr (std::is_same_v<T, std::int8_t>) {
+    return adapter_int8_t{};
+  } else
+#endif
+#ifdef INT16_MAX
+      if constexpr (std::is_same_v<T, std::int16_t>) {
+    return adapter_int16_t{};
+  } else
+#endif
+#ifdef INT32_MAX
+      if constexpr (std::is_same_v<T, std::int32_t>) {
+    return adapter_int32_t{};
+  } else
+#endif
+#ifdef INT64_MAX
+      if constexpr (std::is_same_v<T, std::int64_t>) {
+    return adapter_int64_t{};
+  } else
+#endif
+#ifdef UINT8_MAX
+      if constexpr (std::is_same_v<T, std::uint8_t>) {
+    return adapter_uint8_t{};
+  } else
+#endif
+#ifdef UINT16_MAX
+      if constexpr (std::is_same_v<T, std::uint16_t>) {
+    return adapter_uint16_t{};
+  } else
+#endif
+#ifdef UINT32_MAX
+      if constexpr (std::is_same_v<T, std::uint32_t>) {
+    return adapter_uint32_t{};
+  } else
+#endif
+#ifdef UINT64_MAX
+      if constexpr (std::is_same_v<T, std::uint64_t>) {
+    return adapter_uint64_t{};
+  } else
+#endif
+      if constexpr (std::is_same_v<T, std::byte>) {
+    // std::byte is a scoped enum according to the C++17
+    return adapter_byte{};
+  } else if constexpr (std::is_same_v<T, std::size_t>) {
+    return adapter_size_t{};
+  } else {
+    return adapter_core<T>{};
+  }
+}
+
+};  // namespace details
+
+/** @brief Safe wrapper for types to use within kernel names
+ *  @details According to C++17 [dcl.typedef] type alias does not introduce a
+ *           new type. So we are free to provide adapter for types we cannot use
+ *           for kernel names directly. For example, for any type we cannot
+ *           forward-declare:
+ *               struct outer {
+ *                 struct inner {};
+ *               };
+ *           we are unable to have
+ *               using T = outer::inner;
+ *               single_task<kernel_name<T>>,
+ *           but we are able to have
+ *               using T = typename kernel_name::adapter<outer::inner>::type
```

Comment:
Interesting.
I am curious whether this works on some pure integration-header based implementation like ComputeCpp, because I have the feeling that you need to have `outer::inner` somehow declared in the integration header to be referenced as a template name.
@ProGTX?
@bader I guess in Intel implementation all the useful code might be in the integration header/footer and thus `outer::inner` is found? But in that case, there might not be this constraint on forward-declarability?

Reply:
I'm not sure. We're not able to test this at the moment, but from what I understand the adapter is supposed to work around this limitation?

Reply:
It is unclear whether this should not be forbidden in the specification too. @gmlueck?

Reply:
Yes, I agree that we should clarify the spec about legal kernel names.  I think this CTS code is assuming that the tool which generates the integration header resolves all type aliases and writes the resolved type to the integration header.  To illustrate with a simple example:

```
int main() {
  sycl::queue q;
  using myname = int;

  q.single_task<myname>([=] {/*...*/});
};
```

I think we expect this to work because the integration header should resolve the `myname` alias to `int` and use that in the integration header:

```
template <> struct KernelInfo<int> {/*...*/};
```

Of course, this wouldn't work if the integration header referenced the `myname` alias directly without resolving the type:

```
template <> struct KernelInfo<myname> {/*...*/};  // Doesn't work because "myname" is declared in function scope
```

I think that this CTS code is the same concept, except that the type alias is more complicated.

If others agree with my interpretation, perhaps we need to clarify in the spec that type aliases are allowed as kernel names, and the restrictions we list apply only to the resolved type (not to the alias itself).

Reply:
Just to make sure - the original intent is to have something like this:

```
struct outer {
  struct inner {
  };
};

template <typename T>
struct name {};

int main() {
  sycl::queue q;
  using Foo = typename kernel_name::adapter_t<outer::inner>;
  using Boo = name<kernel_name::adapter_t<std::byte>>;
  q.single_task<Foo>([=] {/*...*/});
  q.single_task<Boo>([=] {/*...*/});
};
```

Reply:
What is the `struct name` used for?